### PR TITLE
#26 Implemented tracking for unsaved changes in the map editor 

### PR DESCRIPTION
**Pull request by Brollyy**
This is to avoid asking user to save when closing the map without any changes.

Separate tracking for info.dat and each difficulty map is implemented due to some actions having partial saving only on info.dat, but not on the difficulty maps.

I've had to add some initialization logic in MainWindow to avoid marking the map as "dirty" when map is first opened or difficulty is changed and the UI elements have their values programatically changed, as this resulted in poor user experience.

### CODE REVIEW

File: Windows/ChangeBPMWindow.xaml.cs [hunk 1]
```diff
@@ -67,6 +67,7 @@ public partial class ChangeBPMWindow : Window {
         }
 
         private void dataBPMChange_CurrentCellChanged(object sender, EventArgs e) {
+            caller.mapEditor.currentMapDifficulty?.MarkDirty(); // This might be too aggressive - triggers even when just clicking on the data cell
```
Comment (PKBeam) [inline Windows/ChangeBPMWindow.xaml.cs:70]: There should be some property in the `EventArgs` that lets you check if the new value is different from the old value. That should prevent the map from being marked dirty with just a click.

Comment (Brollyy) [reply Windows/ChangeBPMWindow.xaml.cs:70]: I'll take a look later today.

File: Windows/MainWindow.xaml.cs [hunk 2]
```diff
@@ -69,6 +69,7 @@ public partial class MainWindow : Window {
         bool shiftKeyDown;
         bool ctrlKeyDown;
         bool returnToStartMenuOnClose = false;
+        bool doneInit = false;
```
Comment (PKBeam) [inline Windows/MainWindow.xaml.cs:72]: I feel like there's a better way we can implement this.

`doneInit` is only set in two functions, right?

Maybe store the dirtiness state of the map before initialisation (i.e. when `doneInit` is set to false), and then afterwards (when `doneInit` is set to true) restore the previous state.

That should stop the map getting marked as dirty by non-user changes.

File: Windows/ChangeBPMWindow.xaml.cs (hunk 1 above)
Comment (Brollyy) [reply Windows/ChangeBPMWindow.xaml.cs]: Reviewing the documentation, this event fires when the cell selection in the grid changes, so it's definitely not the right place to mark as dirty. For edits we already have RowEditEnding, for new rows we have AddingNewItem. One thing that's not covered if I remove it is the BPM change deletion, but I've handled that separately.

File: Windows/MainWindow.xaml.cs (hunk 2 above)
Comment (Brollyy) [reply Windows/MainWindow.xaml.cs]: Done - had to check which state actually was mutated by those initialization to avoid writing too much logic, but it works.

By the way, my change with ```doneInit``` actually had introduced a regression with Recently Viewed songs that I didn't notice before - they weren't updated if you opened a map and immediately closed it, so it's good that this was refactored.

File: Windows/MainWindow.UIControls.cs [hunk 3]
```diff
@@ -381,7 +381,6 @@ public partial class MainWindow : Window {
             recentMaps.RemoveRecentlyOpened(mapEditor.mapFolder);
             recentMaps.AddRecentlyOpened((string)mapEditor.GetMapValue("_songName"), mapEditor.mapFolder);
             recentMaps.Write();
-
```
Comment (Brollyy) [inline Windows/MainWindow.UIControls.cs:384]: Unnecessary change to whitespace

**Review by PKBeam — APPROVED** [review]
